crypto/mlkem/mlkem768.c

```diff
@@ -0,0 +1,1184 @@
+    print_hex(out_shared_secret, OSSL_MLKEM768_SHARED_SECRET_BYTES, "SS2");
+    print_hex(out_ciphertext, OSSL_MLKEM768_CIPHERTEXT_BYTES, "CT2");
+    return 1;
+}
+
+static void decrypt_cpa(uint8_t *out, const struct private_key_RANK768 *priv,
+                        const uint8_t *ciphertext)
+{
+    int du = kDU768;
+    int dv = kDV768;
+    vector u;
+    scalar v, mask;
+
+    vector_decode(&u, ciphertext, du);
+    vector_decompress(&u, du);
+    vector_ntt(&u);
+    scalar_decode(&v, ciphertext + compressed_vector_size(RANK768), dv);
+    scalar_decompress(&v, dv);
+    scalar_inner_product(&mask, &priv->s, &u);
+    scalar_inverse_ntt(&mask);
+    scalar_sub(&v, &mask);
+    scalar_compress(&v, 1);
+    scalar_encode_1(out, &v);
+}
+
+/* See section 6.3 */
+static int mlkem_decap(uint8_t *out_shared_secret,
+                       const uint8_t *ciphertext,
+                       const struct private_key_RANK768 *priv,
+                       ossl_mlkem_ctx *mlkem_ctx)
+{
+    uint8_t decrypted[64];
+    uint8_t key_and_randomness[64];
+    size_t ciphertext_len = ciphertext_size(RANK768);
+    /* TODO(ML-KEM): Maximum also applicable for other algs? */
+    uint8_t expected_ciphertext[OSSL_MLKEM1024_CIPHERTEXT_BYTES];
+    uint8_t failure_key[32];
+    uint8_t mask;
+    int i;
+
+    print_hex((uint8_t *)&priv->pub, sizeof(ossl_mlkem768_public_key), "PK1");
+    print_hex(ciphertext, OSSL_MLKEM768_CIPHERTEXT_BYTES, "CT");
+    decrypt_cpa(decrypted, priv, ciphertext);
+    memcpy(decrypted + 32, priv->pub.public_key_hash,
+           sizeof(decrypted) - 32);
+    hash_g(key_and_randomness, decrypted, sizeof(decrypted), mlkem_ctx);
+    assert(ciphertext_len <= sizeof(expected_ciphertext));
+    encrypt_cpa(expected_ciphertext, &priv->pub, decrypted,
+                key_and_randomness + 32, mlkem_ctx);
+    kdf(failure_key, priv->fo_failure_secret, ciphertext, ciphertext_len, mlkem_ctx);
+    mask = constant_time_eq_int_8(CRYPTO_memcmp(ciphertext,
+                                                expected_ciphertext, ciphertext_len), 0);
+    for (i = 0; i < OSSL_MLKEM768_SHARED_SECRET_BYTES; i++)
+        out_shared_secret[i] = constant_time_select_8(mask,
+                                                      key_and_randomness[i],
+                                                      failure_key[i]);
+    print_hex(out_shared_secret, OSSL_MLKEM768_SHARED_SECRET_BYTES, "SS");
+    return 1;
+}
+
+int ossl_mlkem768_decap(uint8_t *out_shared_secret,
+                        const uint8_t *ciphertext, size_t ciphertext_len,
+                        const ossl_mlkem768_private_key *private_key,
+                        ossl_mlkem_ctx *mlkem_ctx)
+{
+    const struct private_key_RANK768 *priv;
+
+    if (mlkem_ctx == NULL)
+        return 0;
+
+    if (ciphertext_len != OSSL_MLKEM768_CIPHERTEXT_BYTES) {
+        /* TODO(ML-KEM): Review requested randomness strength */
+        RAND_bytes_ex(mlkem_ctx->libctx, out_shared_secret,
```

Comment:
Please check the RAND_bytes_ex return for failure.

Reply:
And what shall we do in this case? This already is the error exit.